experimental/devices/ws2812b/ws2812b.go

```diff
@@ -0,0 +1,506 @@
+}
+
+// raster serializes a buffer of RGB bytes to the WS2812b SPI format.
+//
+// It is expected to be given the part where pixels are, not the header nor
+// footer.
+//
+// dst is in WS2812b SPI 32 bits word format. src is in RGB 24 bits, or 32 bits
+// word format when srcHasAlpha is true. The src alpha channel is ignored in
+// this case.
+//
+// src cannot be longer in pixel count than dst.
+func (d *Dev) raster(dst []byte, src []byte, srcHasAlpha bool) {
+	pBytes := 3
+	if srcHasAlpha {
+		pBytes = 4
+	}
+	length := len(src) / pBytes
+	if l := len(dst) / 4; l < length {
+		length = l
+	}
+	if length == 0 {
+		// Save ourself some unneeded processing.
+		return
+	}
+	stride := 4 //number of spi-bytes in color-byte
+	for i := 0; i < length; i++ {
+		sOff := pBytes * i
+		dOff := 3 * stride * i //9 bytes per LED stride
+		r, g, b := src[sOff], src[sOff+1], src[sOff+2]
+		//grb color order, msb first
+		copy(dst[dOff+stride*0:dOff+stride*1], bitlut[r])
+		copy(dst[dOff+stride*1:dOff+stride*2], bitlut[g])
+		copy(dst[dOff+stride*2:dOff+stride*3], bitlut[b])
+	}
+}
+
+// rasterImg is the generic version of raster that converts an image instead of raw RGB values.
+//
+// It has 'fast paths' for image.RGBA and image.NRGBA that extract and convert the RGB values
+// directly.  For other image types, it converts to image.RGBA and then does the same.  In all
+// cases, alpha values are ignored.
+//
+// rect specifies where into the output buffer to draw.
+//
+// srcR specifies what portion of the source image to use.
+func (d *Dev) rasterImg(dst []byte, rect image.Rectangle, src image.Image, srcR image.Rectangle) {
+	// Render directly into the buffer for maximum performance and to keep
+	// untouched sections intact.
+	switch im := src.(type) {
+	case *image.RGBA:
+		start := im.PixOffset(srcR.Min.X, srcR.Min.Y)
+		// srcR.Min.Y since the output display has only a single column
+		end := im.PixOffset(srcR.Max.X, srcR.Min.Y)
+		// Offset into the output buffer using rect
+		d.raster(dst[4*rect.Min.X:], im.Pix[start:end], true)
+	case *image.NRGBA:
+		// Ignores alpha
+		start := im.PixOffset(srcR.Min.X, srcR.Min.Y)
+		// srcR.Min.Y since the output display has only a single column
+		end := im.PixOffset(srcR.Max.X, srcR.Min.Y)
+		// Offset into the output buffer using rect
+		d.raster(dst[4*rect.Min.X:], im.Pix[start:end], true)
+	default:
+		// Slow path.  Convert to RGBA
+		b := im.Bounds()
+		m := image.NewRGBA(image.Rect(0, 0, b.Dx(), b.Dy()))
+		draw.Draw(m, m.Bounds(), src, b.Min, draw.Src)
+		start := m.PixOffset(srcR.Min.X, srcR.Min.Y)
+		// srcR.Min.Y since the output display has only a single column
+		end := m.PixOffset(srcR.Max.X, srcR.Min.Y)
+		// Offset into the output buffer using rect
+		d.raster(dst[4*rect.Min.X:], m.Pix[start:end], true)
+	}
+}
+
+var _ display.Drawer = &Dev{}
+
+// The bit lookup table converts a single byte into its 3 byte SPI symbol
+// 0 => 1000 and 1 => 1110
+var bitlut = map[byte][]byte{
```

Comment:
> A map[byte] is the equivalent of [256]. An array is more efficient for both memory usage and access speed. 
The map (and the lookup table as a whole) is a byproduct of hacking this to a working state- resource concerns took a back seat to debug legibility. Happy to move this to a [256]byte.

> [NRZ func] is fast enough so [the lookup table (?)] is not needed
The mapping is static, so calling a translation function on each write seems dubious. Even if it's fast enough, it's work we don't really need to do. Of course, a giant hardcoded lookup table is odd too. What if we used NRZ or similar to generate the lookup table during the proposed call to nrzled.NewSPI()? 

Thanks for the review! Expect a commit to this branch incorporating these discussions in the next few days.

Reply:
Fine with hardcoding the table, as long as there's a unit test that confirms that the values are the same with the algorithm. This can be done as a separate CL.

Sorry for requesting big changes, and thanks for sending this PR. I think this is a cool idea.

Eventually the driver should be smart about the SPI buffer size (i.e. https://periph.io/platform/raspberrypi/#spi-buffer-size) and at least when possible adjust and error accordingly. It can be done by querying if the spi.Port implements https://periph.io/x/periph/host/sysfs#SPI.MaxTxSize